internal/core/software_renderer.rs

```diff
@@ -1667,6 +1731,18 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
                 self.draw_text_paragraph(paragraph, physical_clip, offset, color);
             }
         }
+
+        let cursor_rect = self
+            .window
+            .window_adapter()
+            .renderer()
+            .text_input_cursor_rect_for_byte_offset(
+                text_input,
+                text_input.cursor_position_byte_offset().round() as usize,
+            )
+            .cast_unit::<crate::lengths::PhysicalPx>();
+        self.processor
+            .process_rectangle(cursor_rect.translate(offset.cast()).cast::<i16>(), color.into());
```

Comment:
This kind of works. It's not correct for cursor placement in password input fields though. The entire draw function should fetch the visual representation once and render it and draw the cursor (if necessary).

The thing about this `TextVisualRepresentation` is that it covers a few things:

1) It takes care of determining if the cursor needs to be shown or not.
2) It computes the correct text selection byte offsets (swapping anchor and position if necessary).
3) It replaces the text with the password character if necessary, and adjusts the byte offsets for the cursor accordingly. It may happen for example that an ascii character is replaced with a multi-byte (utf-8) password character - or the other way around, in which case the byte offsets for the cursor and selection need to be adjusted.



```suggestion
        if let Some(cursor_rect) =
            text_input.visual_representation(None).cursor_position.map(|pos| {
                self.window
                    .window_adapter()
                    .renderer()
                    .text_input_cursor_rect_for_byte_offset(text_input, pos)
                    .cast()
                    * self.scale_factor
            })
        {
            self.processor.process_rectangle(cursor_rect.translate(offset).cast(), color.into());
        }
```